### ISSUE TITLE
Error on plugin load: runtime.tlsg: initial-exec TLS resolves to dynamic definition

### ISSUE THREAD

**Issue by huiser** [iegomez/mosquitto-go-auth#20]
2019-05-29T16:26:23: mosquitto version 1.6.2 starting
2019-05-29T16:26:23: Config loaded from /mosquitto/mosquitto.conf.
2019-05-29T16:26:23: Loading plugin: /mosquitto/go-auth.so
2019-05-29T16:26:23: Error: Unable to load auth plugin "/mosquitto/go-auth.so".
2019-05-29T16:26:23: Load error: Error relocating /mosquitto/go-auth.so: runtime.tlsg: initial-exec TLS resolves to dynamic definition in /mosquitto/go-auth.so

Mosquitto is built from source in Alpine Linux Docker container

**Comment by huiser** [comment]
# Builder image
FROM golang:1.12-alpine as builder

ENV PATH=/src/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

RUN install -d /src && \
    install -d /src/mosquitto && \
    install -d /src/mosquitto-auth

RUN apk add --no-cache \
    curl \
    c-ares-dev \
    gcc \
    git \
    g++ \
    libwebsockets-dev \
    make \
    musl-dev \
    openssl-dev \
    sed \
    tar

RUN curl -s https://mosquitto.org/files/source/mosquitto-1.6.2.tar.gz | tar --strip-components=1 -C /src/mosquitto -xzf -
RUN make -C /src/mosquitto \
    WITH_MEMORY_TRACKING=no \
    WITH_WEBSOCKETS=yes \
    WITH_SRV=yes \
    WITH_TLS_PSK=no \
    WITH_ADNS=no \
    prefix=/usr \
    binary && \
    make -C /src/mosquitto \
    prefix=/usr \
    DESTDIR="/install" \
    install

RUN curl -s -L https://github.com/iegomez/mosquitto-go-auth/archive/0.3.0.tar.gz | tar --strip-components=1 -C /src/mosquitto-auth -xzf -
RUN cd /src/mosquitto-auth && \
    CGO_CFLAGS="-I/install/usr/include -fPIC" \
    CGO_LDFLAGS="-shared" \
    make

# not used:
    #CGO_ENABLED="1" \
    #GOOS="linux" \

# Final image
FROM alpine
LABEL description="Eclipse Mosquitto MQTT Broker"
ENV USER=mosquitto \
    UID=1000 \
    GID=1000

RUN addgroup --gid "$GID" "$USER" \
    && adduser \
    --disabled-password \
    --gecos "" \
    --home "/mosquitto" \
    --ingroup "$USER" \
    --no-create-home \
    --uid "$UID" \
    "$USER"

RUN install -d -o mosquitto /mosquitto && \
    install -d -o mosquitto /mosquitto/config && \
    install -d -o mosquitto /mosquitto/data && \
    install -d -o mosquitto /mosquitto/log

RUN apk add --no-cache \ 
    c-ares \
    libwebsockets \
     musl

COPY --from=builder /install /
COPY --from=builder /src/mosquitto-auth/go-auth.so /mosquitto/

VOLUME /mosquitto/config /mosquitto/data /mosquitto/log

WORKDIR /mosquitto

USER mosquitto

EXPOSE 1883 8883

COPY mosquitto.conf /mosquitto/

ENTRYPOINT ["mosquitto"]
CMD ["-c", "/mosquitto/mosquitto.conf"]

**Comment by huiser** [comment]
[Dockerfile.txt](https://github.com/iegomez/mosquitto-go-auth/files/3233510/Dockerfile.txt)

**Comment by iegomez** [comment]
There's an [ongoing bug](https://github.com/golang/go/issues/13492) when using cgo (which the plugin uses to implement mosquitto's expected plugin interface) on Alpine as it uses muslibc instead of glibc, as referenced in https://github.com/iegomez/mosquitto-go-auth/issues/14.

I mentioned some possible workarounds at https://github.com/iegomez/mosquitto-go-auth/issues/15#issuecomment-478582777, but still haven't tried any of those, so it'd be really helpful if you are willing and have the time to give it a go.

**Comment by huiser** [comment]
I can't get it working with Alpine, using the suggestion above.
I switched to the 'debian:stable-slim' base image, for now. If I have time, I will give it another try.

**Comment by iegomez** [comment]
Ok, thanks. I don't use docker and don't have the time right now to learn it properly and come up with a decent image, but I'll leave this open for the future.

In case you get it to work or believe that an alternative such as using `debian:stable-slim` is good enough and you want to contribute with docker related additions just make a PR, or let me know if you'd like me to link to an external repo with instructions on how to run a docker image with mosquitto and the plugin in the documentation.

**Comment by iegomez** [comment]
Hi, @huiser. I tried some workarounds a couple of days ago but couldn't get it to work with Alpine. I had started building another image based on `debian:stable-slim` but we got robbed at work and hadn't pushed it.

So I intend to rework on it next monday and add it to the repo in a `docker` directory that would look something like this:
```
-mosquitto-go-atuh
  -docker
    -auth
      acls
      passwords
  Dockerfile
  mosquitto.conf
  go-auth.conf
```

That's unless you wish to contribute your `debian:stable-slim` image. Just let me know if that's the case.
